flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java

```diff
@@ -0,0 +1,179 @@
+		} else {
+			commands.addAll(Arrays.asList("--install-option", "--prefix=" + requirementsInstallDir));
+		}
+		if (requirementsCacheDir != null) {
+			commands.addAll(Arrays.asList("--find-links", requirementsCacheDir));
+		}
+
+		int retries = 0;
+		while (true) {
+			try {
+				executeHelperScript(commands.toArray(new String[0]), environmentVariables);
+				break;
+			} catch (Throwable t) {
+				retries++;
+				if (retries < MAX_RETRY_TIMES) {
+					LOG.warn(String.format("Pip install failed, retrying... (%d/%d)", retries, MAX_RETRY_TIMES), t);
+				} else {
+					LOG.error(String.format("Pip install failed, already retried %d time...", retries));
+					throw t;
+				}
+			}
+		}
+	}
+
+	public static String getRunnerScriptPath(
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String[] commands = new String[] { pythonExecutable, "-c", GET_RUNNER_SCRIPT_DIR };
+		String out = executeHelperScript(commands, environmentVariables);
+		String runnerScriptPath;
+		if (OperatingSystem.isWindows()) {
+			runnerScriptPath = String.join(File.separator, out.trim(), PYFLINK_UDF_RUNNER_BAT);
+		} else {
+			runnerScriptPath = String.join(File.separator, out.trim(), PYFLINK_UDF_RUNNER_SH);
+		}
+		if (!new File(runnerScriptPath).exists()) {
+			throw new FileNotFoundException(String.format(
+				"The runner script '%s' does not exist! " +
+				"Please reinstall the apache-flink Python package.", runnerScriptPath));
+		}
+		return runnerScriptPath;
+	}
+
+	private static String getSitePackagesPath(
+			String prefix,
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String[] commands = new String[] { pythonExecutable, "-c", GET_SITE_PACKAGES_PATH_SCRIPT, prefix };
+		String out = executeHelperScript(commands, environmentVariables);
+		return String.join(File.pathSeparator, out.trim().split("\n"));
+	}
+
+	private static boolean isPipVersionNotLowerThan(
+			String pipVersion,
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String[] commands = new String[] { pythonExecutable, "-c", CHECK_PIP_VERSION_SCRIPT, pipVersion };
+		String out = executeHelperScript(commands, environmentVariables);
+		return Boolean.valueOf(out.trim());
+	}
+
+	private static String executeHelperScript(
+			String[] commands,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		ProcessBuilder pb = new ProcessBuilder(commands);
+		pb.environment().putAll(environmentVariables);
+		pb.redirectErrorStream(true);
+		Process p = pb.start();
+		InputStream in = new BufferedInputStream(p.getInputStream());
+		BufferedReader br = new BufferedReader(new InputStreamReader(in));
+		StringBuilder out = new StringBuilder();
+		String s;
+		while ((s = br.readLine()) != null) {
+			out.append(s).append("\n");
+		}
+		if (p.waitFor() != 0) {
```

Comment:
I think it is necessary to use `waitFor(long timeout, TimeUnit unit)` as the subprocess is already exited when the `br.readLine` return `null`. I agree that the InterruptedException could be ignored because the `waitFor` method would not be blocked in this situation.